C/STrace/EtwLogger.cpp

```diff
@@ -0,0 +1,460 @@
+	const auto dataDescriptors = (PEVENT_DATA_DESCRIPTOR)ExAllocatePoolWithTag(NonPagedPoolNx, allocSize, 'wteP');
+	if (dataDescriptors == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	memset(dataDescriptors, 0, allocSize);
+
+	// Set the provider and event metadata descriptors.
+	dataDescriptors[0] = m_providerMetadataDesc;
+	dataDescriptors[1] = event->MetadataDescriptor();
+
+	// Create a descriptor for each individual field.
+	va_list args;
+	va_copy(args, fields);
+	for (auto i = 0; i < numberOfFields; i++)
+	{
+		va_arg(args, const char*);
+		const auto fieldType = va_arg(args, int);
+		auto fieldValue = va_arg(args, size_t);
+
+		dataDescriptors[i + 2] = CreateTraceProperty(
+			fieldType,
+			fieldType != EtwFieldAnsiString ? &fieldValue : (void*)fieldValue);
+		if (dataDescriptors[i + 2].Ptr == NULL)
+		{
+			return STATUS_UNSUCCESSFUL;
+		}
+	}
+	va_end(args);
+
+	// Create the event descriptor.
+	EVENT_DESCRIPTOR eventDescriptor;
+	memset(&eventDescriptor, 0, sizeof(EVENT_DESCRIPTOR));
+	eventDescriptor.Channel = 11;  // All "manifest-free" events should go to channel 11 by default
+	eventDescriptor.Keyword = keyword;
+	eventDescriptor.Level = eventLevel;
+
+	// Write the event.
+	return EtwWrite(m_regHandle, &eventDescriptor, NULL, numberOfDescriptors, dataDescriptors);
+
+	// TODO: This leaks the memory allocated for the field descriptors
+}
+
+LPCGUID EtwProvider::Guid() const noexcept
+{
+	return m_guid;
+}
+
+EtwProviderEvent* EtwProvider::FindEvent(const char* eventName)
+{
+	for (auto i = 0; i < m_events.len(); i++)
+	{
+		if (strcmp(eventName, m_events[i].Name()) == 0)
+		{
+			return &m_events[i];
+		}
+	}
+
+	return NULL;
+}
+
+size_t EtwProvider::SizeOfField(int fieldType, void* fieldValue)
+{
+	size_t sizeOfField = 0;
+
+	switch (fieldType & 0x000000FF)
+	{
+	case EtwFieldUnicodeString:
+		sizeOfField = (wcslen((wchar_t*)fieldValue) + 1) * sizeof(wchar_t);
+		break;
+	case EtwFieldAnsiString:
+		sizeOfField = strlen((char*)fieldValue) + 1;
+		break;
+	case EtwFieldInt8:
+		sizeOfField = sizeof(int8_t);
+		break;
+	case EtwFieldUInt8:
+		sizeOfField = sizeof(uint8_t);
+		break;
+	case EtwFieldInt16:
+		sizeOfField = sizeof(int16_t);
+		break;
+	case EtwFieldUInt16:
+		sizeOfField = sizeof(uint16_t);
+		break;
+	case EtwFieldInt32:
+		sizeOfField = sizeof(int32_t);
+		break;
+	case EtwFieldUInt32:
+		sizeOfField = sizeof(uint32_t);
+		break;
+	case EtwFieldInt64:
+		sizeOfField = sizeof(int64_t);
+		break;
+	case EtwFieldUInt64:
+		sizeOfField = sizeof(uint64_t);
+		break;
+	case EtwFieldFloat:
+		sizeOfField = sizeof(float);
+		break;
+	case EtwFieldDouble:
+		sizeOfField = sizeof(double);
+		break;
+	case EtwFieldBool32:
+		sizeOfField = sizeof(int32_t);
+		break;
+	case EtwFieldGuid:
+		sizeOfField = sizeof(GUID);
+		break;
+	case EtwFieldHexInt32:
+		sizeOfField = sizeof(int32_t);
+		break;
+	case EtwFieldHexInt64:
+		sizeOfField = sizeof(int64_t);
+		break;
+	// TODO: more fields
```

Comment:
Let's clean this up, or comment which ones we don't implement (if there's not too many).

Reply:
Just looked at this, and I'd missed a bit from the implementation - there was an additional byte for the OutType on some fields. I think I've got this down, the TODO is removed, though the switch statement is even bigger now.